workbench-core/environments/src/handlers/environmentTypeHandler.ts

```diff
@@ -0,0 +1,127 @@
+    this._mainAccountAwsService = mainAccountAwsService;
+  }
+  /* eslint-disable-next-line @typescript-eslint/no-explicit-any*/
+  public async execute(event: any): Promise<void> {
+    console.log(`Processing Environment Types`);
+    const portfolioName = process.env.SC_PORTFOLIO_NAME!;
+    const portfolioId = await this._mainAccountAwsService.helpers.serviceCatalog.getPortfolioId(
+      portfolioName
+    );
+    if (portfolioId === undefined) {
+      throw new Error(`Could not find portfolioId for portfolio: ${portfolioName}`);
+    }
+    console.log(`Processing portfolio ${portfolioId}`);
+    const products = await this._mainAccountAwsService.helpers.serviceCatalog.getProductsByPortfolioId(
+      portfolioId
+    );
+    for (const product of products) {
+      try {
+        console.log(`Processing product ${product.ProductId}`);
+        const currentProductId = product.ProductId || '';
+        const provisionArtifacts =
+          await this._mainAccountAwsService.helpers.serviceCatalog.getProvisionArtifactsByProductId(
+            currentProductId
+          );
+        for (const provisionArtifact of provisionArtifacts.filter((p) => p.Active)) {
+          try {
+            console.log(`Processing provision artifact ${provisionArtifact.Id}-${product.ProductId}`);
+            const details =
+              await this._mainAccountAwsService.helpers.serviceCatalog.getProvisionArtifactDetails(
+                currentProductId,
+                provisionArtifact.Id || ''
+              );
+            const envTypeId = `et-${product.ProductId},${provisionArtifact.Id}`;
+            const existingEnvType = await this._getExistingEnvironmentType(envTypeId);
+            if (!existingEnvType && !!details?.Info?.TemplateUrl) {
+              console.log(`Creating environment type: ${envTypeId}.`);
+              const params = await this._mainAccountAwsService.helpers.s3.getTemplateByURL(
+                details.Info.TemplateUrl
+              );
+              await this._saveEnvironmentType(product, provisionArtifact, params as ProvisionArtifactParams);
+            }
+          } catch (e) {
+            console.log(
+              `An error ocurred while trying to process Provision Artifact: ${provisionArtifact.Id} - ${e}`
+            ); //continue process when a provision artifact fails
+          }
+        }
+      } catch (e) {
+        console.log(`An error ocurred while trying to process Product: ${product.ProductId} - ${e}`); //continue process when a product fails
+      }
+    }
+  }
+
+  private async _saveEnvironmentType(
+    product: ProductViewSummary,
+    provisionArtifact: ProvisioningArtifactDetail,
+    provisionParams: ProvisionArtifactParams
+  ): Promise<void> {
+    const envTypeService = new EnvironmentTypeService({ TABLE_NAME: process.env.STACK_NAME! });
+    await envTypeService.createNewEnvironmentType({
+      productId: product.ProductId || '',
```

Comment:
I think we should add a code segment above this to confirm that a valid `ProductId` and `provisioningArtifact.Id`. If those values don't exist we should throw an error. If those values are saved as empty string it's going to cause errors later on.

Reply:
validation added

Reply:
Since there is already a check at line 75 can we structure the code so the productId being passed in is a verified `string`? We don't need to do `product.ProductId || ''` anymore. Maybe something like this.

```
let productId: string = '';
if (!product?.ProductId || !provisionArtifact?.Id) {
      throw new Error(
        `An error ocurred while saving an Environment Type, Product and Artifact Must not be empty, { Product: '${
          product?.ProductId ?? ''
        }', Provision Artifact: '${provisionArtifact?.Id ?? ''}'}`
} else {
    productId = product.ProductId!;
    // do the same for provisionArtifact 
}

await envTypeService.createNewEnvironmentType({
      productId,
      ...

```

Reply:
Typescript seems to be smart enough to know that after that if product and pa will not be undefined , so I only removed the null check